### PR TITLE
Add opt-in access control for infrastructure operators

### PR DESCRIPTION
**Pull request by DanGould**
This came straight from the trough but I don't think it's actually slop.

I had opus 4.6 draft this (much faster turnover, better integration with GitHub apparently) and codex 5.3 review it and then did a bunch of manual review with the help of codex  to get CI passing and understand the design tradeoffs.. nix flaking took forever (1h+), maybe because I just reinstalled nix

## Summary

Adds a feature-gated (`access-control`) opt-in system for infrastructure operators to control access to their payjoin-mailroom instances. All behavior is off by default and requires explicit configuration.

- **GeoIP region filtering** — middleware blocks connections from configurable ISO 3166-1 alpha-2 regions. Auto-fetches DB-IP Lite (CC-BY 4.0) if no custom MMDB path is provided.
- **V1 address screening** — parses V1 PSBTs and checks output/input addresses against a configurable blocklist file.
- **Address list auto-fetch** — background task fetches blocklist from a configurable URL on a refresh interval, with disk cache fallback.
- **V1 enable toggle** — config flag to accept all V1 requests otherwise rejected.
- **V1 data minimization** — clears plaintext V1 payload from memory immediately after first read.

## Test plan

- [x] 21 tests passing across `payjoin-mailroom` and `payjoin-directory`
- [x] Compiles cleanly with and without `access-control` feature flag
- [x] `cargo clippy --all-features` clean
- [ ] Manual testing with a real GeoIP database and blocklist
- [x] Review test MMDB fixture (`test-data/GeoIP2-Country-Test.mmdb`) is appropriate for CI

### CODE REVIEW

File: payjoin-mailroom/src/access_control.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+use std::net::IpAddr;
+use std::path::Path;
+
+use maxminddb::PathElement;
+
+use crate::config::AccessControlConfig;
+
+pub struct AccessControl {
+    geo_reader: Option<maxminddb::Reader<Vec<u8>>>,
+    blocked_regions: HashSet<String>,
```
Comment (spacebear21) [inline payjoin-mailroom/src/access_control.rs:11]: It seems natural that it should also have a custom `blocked_ips` HashSet for custom IP ranges.

File: payjoin-mailroom/src/access_control.rs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+impl AccessControl {
+    pub async fn from_config(
+        config: &AccessControlConfig,
+        storage_dir: &Path,
+    ) -> anyhow::Result<Self> {
+        let geo_reader = match &config.geo_db_path {
+            Some(path) => Some(maxminddb::Reader::open_readfile(path)?),
+            None if !config.blocked_regions.is_empty() => {
+                let cached = storage_dir.join("geoip.mmdb");
```
Comment (spacebear21) [inline payjoin-mailroom/src/access_control.rs:22]: nit: if other access control things will be stored (e.g. address blocklists), we should group them all in a access control subdir

File: payjoin-mailroom/src/access_control.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    let era = if z >= 0 { z } else { z - 146_096 } / 146_097;
+    let doe = z - era * 146_097;
+    let yoe = (doe - doe / 1_460 + doe / 36_524 - doe / 146_096) / 365;
+    let y = yoe + era * 400;
+    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
+    let mp = (5 * doy + 2) / 153;
+    let month = (mp + if mp < 10 { 3 } else { -9 }) as u32;
+    let year = (y + if month <= 2 { 1 } else { 0 }) as i32;
+    (year, month)
+}
```
Comment (spacebear21) [inline payjoin-mailroom/src/access_control.rs:122]: This looks like LLM insanity, does std::time not have a utility that could achieve the same effect? Looks like we just need the current year and month?

File: flake.nix
Comment (spacebear21) [inline flake.nix:166]: We need to enable the `access-control` feature here if we want it to be accessible in the nix flake and docker images built with nix2container.

File: payjoin-directory/src/lib.rs [hunk 4]
```diff
@@ -375,6 +383,16 @@ impl<D: Db> Service<D> {
         B::Error: Into<BoxError>,
     {
         trace!("Post fallback v1");
+
+        if self.v1_disabled {
+            return Ok(Response::builder()
+                .status(StatusCode::FORBIDDEN)
+                .header(CONTENT_TYPE, "application/json")
+                .body(full(
+                    r#"{"errorCode": "v1-disabled", "message": "V1 is disabled on this server"}"#,
```
Comment (spacebear21) [inline payjoin-directory/src/lib.rs:392]: Should this use the `version-unsupported` BIP78 error code? Per the BIP "it is important that error codes that are not well-known and that the message do not appear on the sender's software user interface"

File: payjoin-directory/src/lib.rs [hunk 5]
```diff
@@ -438,6 +474,84 @@ impl<D: Db> Service<D> {
...
+    ParseError(String),
+}
+
+fn screen_v1_addresses(body: &str, blocked: &std::collections::HashSet<String>) -> ScreenResult {
+    use bitcoin::base64::prelude::{Engine, BASE64_STANDARD};
+    use bitcoin::psbt::Psbt;
+    use bitcoin::{Address, Network};
+
+    // V1 payjoin deployments in use today are Bitcoin mainnet-only.
+    let network = Network::Bitcoin;
```
Comment (spacebear21) [inline payjoin-directory/src/lib.rs:489]: Does this prevent e.g. regtest backwards compatibility testing, presuming the decoding would always fail?

File: payjoin-mailroom/src/lib.rs [hunk 6]
```diff
@@ -38,9 +38,13 @@ pub async fn serve(config: Config, meter_provider: Option<SdkMeterProvider>) ->
 
     #[cfg(feature = "access-control")]
     let access_control = init_access_control(&config).await?;
+    #[cfg(feature = "access-control")]
+    let blocked_addresses = init_blocked_addresses(&config)?;
+    #[cfg(not(feature = "access-control"))]
+    let blocked_addresses = None;
```
Comment (spacebear21) [inline payjoin-mailroom/src/lib.rs:44]: Seems like these should be a part of `init_access_control`

File: payjoin-mailroom/src/lib.rs [hunk 7]
```diff
@@ -23,18 +29,27 @@ struct Services {
     directory: payjoin_directory::Service<payjoin_directory::FilesDb>,
     relay: ohttp_relay::Service,
     metrics: MetricsService,
+    #[cfg(feature = "access-control")]
+    access_control: Option<std::sync::Arc<access_control::AccessControl>>,
```
Comment (spacebear21) [inline payjoin-mailroom/src/lib.rs:33]: Why does this need to be an `Arc<>`?

File: payjoin-directory/src/lib.rs [hunk 8]
```diff
@@ -68,6 +68,8 @@ pub struct Service<D: Db> {
     db: D,
     ohttp: ohttp::Server,
     sentinel_tag: SentinelTag,
+    blocked_addresses: Option<Arc<tokio::sync::RwLock<std::collections::HashSet<String>>>>,
```
Comment (spacebear21) [inline payjoin-directory/src/lib.rs:71]: Is the `Arc<tokio::sync::RwLock<>>` really necessary?

File: payjoin-mailroom/src/access_control.rs [hunk 9]
```diff
@@ -71,6 +71,28 @@ impl AccessControl {
...
+    } else {
+        addr.to_string()
+    }
+}
+
+fn is_bech32_address(addr: &str) -> bool {
+    let lower = addr.to_ascii_lowercase();
+    lower.starts_with("bc1") || lower.starts_with("tb1") || lower.starts_with("bcrt1")
+}
+
```
Comment (spacebear21) [inline payjoin-mailroom/src/access_control.rs:95]: Why not parse these as `bitcoin::Address` here on read, and let rust-bitcoin handle weird bitcoin things on parsing and equality checks?

File: payjoin-mailroom/src/access_control.rs [hunk 10]
```diff
@@ -71,6 +71,28 @@ impl AccessControl {
     }
 }
 
+pub fn load_blocked_addresses(path: &Path) -> anyhow::Result<HashSet<String>> {
```
Comment (spacebear21) [inline payjoin-mailroom/src/access_control.rs:74]: Shouldn't `AccessControl` be in charge of loading and checking blocked addresses? Are they only separate to accommodate `payjoin-directory` still living in a separate crate?

File: payjoin-mailroom/src/lib.rs [hunk 11]
```diff
@@ -247,18 +247,94 @@ async fn init_access_control(
...
+            },
+            Ok(resp) => {
+                tracing::warn!("Failed to fetch address list: HTTP {}", resp.status());
+                try_load_cache(&cache_path, &blocked).await;
+            }
+            Err(e) => {
+                tracing::warn!("Failed to fetch address list: {e}");
+                try_load_cache(&cache_path, &blocked).await;
+            }
+        }
```
Comment (spacebear21) [inline payjoin-mailroom/src/lib.rs:311]: error handling/tracing here seems excessive

File: payjoin-mailroom/src/lib.rs [hunk 12]
```diff
@@ -337,6 +355,18 @@ async fn try_load_cache(
     }
 }
 
+fn get_v1_disabled(config: &Config) -> bool {
```
Comment (spacebear21) [inline payjoin-mailroom/src/lib.rs:358]: Why even have this, it's just flipping the `enable_v1` config value. init_directory already takes a `config`, no need to pass a separate v1_disabled parameter

File: payjoin-mailroom/src/config.rs [hunk 13]
```diff
@@ -63,6 +63,7 @@ pub struct AccessControlConfig {
     pub blocked_addresses_path: Option<PathBuf>,
     pub blocked_addresses_url: Option<String>,
     pub blocked_addresses_refresh_secs: Option<u64>,
+    pub enable_v1: bool,
```
Comment (spacebear21) [inline payjoin-mailroom/src/config.rs:66]: I'm not sure this should be scoped to `AccessControlConfig`, it's not strictly access control, an operator might simply not want to store plaintext. Also it's confusing because currently v1 is enabled by default, unless `--features=access-control` in which case it's disabled by default and needs to be re-enabled explicitly.

File: payjoin-mailroom/src/access_control.rs (hunk 1 above)
Comment (DanGould) [reply payjoin-mailroom/src/access_control.rs:11]: Agree. Nothing stopping us from shipping that. This is an absolute bare minimum off the shelf strategy to work with easy to access DBs

File: payjoin-mailroom/src/access_control.rs (hunk 3 above)
Comment (DanGould) [reply payjoin-mailroom/src/access_control.rs:122]: std::time does not have a utility to achieve the same affect, we'd need to introduce `chrono` or `time`.

Yes, this was plainly and obviously LLM written but also checked thereafter and recognized that it's used to fetch the updated DB on a schedule.  Kind of like how wikipedia is really good for technical / algorithmic explanation, so is LLM for concrete well understood algorithms and fetch unit tests for well-known time edge cases. This algo is used to fetch the date to get an updated DB at the right time, I think it's acceptable to take this rather than a dependency for that purpose.

File: payjoin-directory/src/lib.rs (hunk 4 above)
Comment (DanGould) [reply payjoin-directory/src/lib.rs:392]: Yes definitely, if not the UI will just show a generic error message

File: payjoin-directory/src/lib.rs (hunk 5 above)
Comment (DanGould) [reply payjoin-directory/src/lib.rs:489]: No, this is fail-safe. Regtest string address would not match a bitcoin string and then not be blocked ever.

File: payjoin-mailroom/src/lib.rs (hunk 6 above)
Comment (DanGould) [reply payjoin-mailroom/src/lib.rs:44]: Then init_access_contrrol just needs a rename, because that's IP regions which the whole mailroom needs and this is addresses, which only the directory needs.

File: payjoin-mailroom/src/lib.rs (hunk 7 above)
Comment (DanGould) [reply payjoin-mailroom/src/lib.rs:33]: Yes. Did you try yourself to rip it out or ask the robot to rip it out and see what happens?

the same `AccessControl` value must be shared as part of services. 
` .with_state(services)` so Axum can give handlers access to it.

see:
```rs
router = router
            .layer(axum::middleware::from_fn(middleware::check_access_control))
            .layer(axum::Extension(acaccess_control));
```

Without `Arc` it'd need to be `Clone`. `maxminddb::Reader<Vec<u8>>` is a field of `AccessControl` and is itself not clone. We also need this across multiple threads.

See the mailboxes FileDb and ohttp_relay::Service for the same arch.

```rs
pub struct Db {
    timeout: Duration,
    mailboxes: Arc<Mutex<Mailboxes>>,
}
```

The only thing I might change would be to encapsulated the `Arc` inside the `AccessControl` struct

File: payjoin-directory/src/lib.rs (hunk 8 above)
Comment (DanGould) [reply payjoin-directory/src/lib.rs:71]: Yes, `Arc` for similar reasons to `AccessControl` above: we need to share ownership / cheap clone this during concurrent requests / in multiple paces at once / across threads. `RwLock` because we're writing at runtime when a timer triggers the update of blocked addresses.

File: payjoin-mailroom/src/access_control.rs (hunk 10 above)
Comment (DanGould) [reply payjoin-mailroom/src/access_control.rs:74]: No, AccessControl is the Region-specific control, relevant to the top-level mailroom. Blocked addresses indeed are a separate struct / update path to be passed specifically to payjoin-directory as a separate module & lower level of abstraction, whether it's in a separate crate or not doesn't matter here.

Comment (DanGould) [reply payjoin-mailroom/src/access_control.rs:74]: It could be made to encapsulate both geo and blocked addresses but right now they're separate

File: payjoin-mailroom/src/lib.rs (hunk 11 above)
Comment (DanGould) [reply payjoin-mailroom/src/lib.rs:311]: It'd help me correct this if you'd define why you think this is excessive, and what level of granularity you'd accept. These are errors on once every 24h checks, and if they fail it'd be good to know precisely why

File: payjoin-mailroom/src/config.rs (hunk 13 above)
Comment (DanGould) [reply payjoin-mailroom/src/config.rs:66]: I agree this config isn't quite right

File: payjoin-mailroom/src/access_control.rs (hunk 9 above)
Comment (DanGould) [reply payjoin-mailroom/src/access_control.rs:95]: Because it's at the mailroom rather than the directory and we're not parsing addresses we're just matching the strings. I can instead have the mailroom do the fetching and just have the directory load the addresses from a path and do loading and normalizing, but I do think this would have worked as-is without giving us a ton of trouble.

**Comment by DanGould** [comment]
> It's definitely way sloppier than your usual code quality 🥲 I reviewed the first four commits and found a lot to be desired, posting what I have so far

I agree, and there's a conversation to be had about the acceptable level of sloppiness for something that's 1. on the periphery of our integrations rather than library core 2. a burning time crunch blocking us from proceeding elsewhere.

Your comments about 

1. the error message for v1 
2. Features necessary for flake builds

do need to be addressed here asap to ship. The other comment about

- the way the configuration of v1 ability could be better

is noted.

Can you be specific about what's left to be desired? This kind of change seems largely mechanical and peripheral to our main integration goals, which is why I felt a whole week of my time focused on this was not worth it compared to shipping something that works and isn't broken that some folks with less experience and more free cycles can iterate on and refine (as they already are doing)
